### PR TITLE
ci: attempt to persist to workspace

### PR DESCRIPTION
**Pull request by Westbrook**
This PR is a work in progress of a proof of concept of using CircleCI workflows to decrease build times.

### CODE REVIEW

**Comment by LarsDenBakker** [comment]
For the "fast lane" we could probably run browser tests in local chrome

**Comment by Westbrook** [comment]
@LarsDenBakker I think the "fast lane" is getting to be in a good place here. Thoughts?

**Comment by Westbrook** [comment]
I don't have access to the `Settings` page for this repo, but with this update, we'd be able to change the branch protection for `master` so that only `build` and `test-local` are required for the PRs to go green.

**Comment by LarsDenBakker** [comment]
Looks good to me, when will we run the slow lane then?

**Comment by Westbrook** [comment]
Right now I’m always running the slow lane. It’s not taking that much longer than the fast lane anyways, the idea is you wouldn’t block PRs on the slow lane, just the fast one.

Who has access to tweak the branch rules?

**Comment by LarsDenBakker** [comment]
ah, that makes perfect sense! I think only @daKmoR has this access on the repository

**Comment by Westbrook** [comment]
@daKmoR, would love to chat through any issues you might have here so we can get the branch settings updated to let this go green and move forward with it. Let me know what works best for you!

File: .circleci/config.yml [hunk 1]
```diff
@@ -40,23 +31,48 @@ jobs:
       # auto build typescript definition files (not part of build as locally JsDoc alone is all we need)
       - run: npm run build:types
 
+      # persist built files and duplicate dependencies to workspace
+      - persist_to_workspace:
+          root: ~/
+          paths:
+              - repo/packages/*/node_modules
+              - repo/packages/*/dist
```
Comment (daKmoR) [inline .circleci/config.yml]: should we add `- repo/node_modules` as well?

File: .circleci/config.yml [hunk 2]
```diff
@@ -73,10 +89,17 @@ workflows:
...
   test-deploy:
     jobs:
-      - test
+      - build
+      - test-local:
+          requires:
+            - build
+      - test-bs:
+          requires:
+            - build
```
Comment (daKmoR) [inline .circleci/config.yml]: I think we should add `- test-local` here as well?

e.g. we only wanna bother browserstack once the local test succeeded right?

File: .circleci/config.yml (hunk 1 above)
Comment (Westbrook) [reply .circleci/config.yml]: We could do a couple of test builds to compare performance, but right now they are being added to the cache, which can persist longer (as long as yarn.lock doesn’t change) but can’t be targeted by glob syntax. Being `node_modules` can be rather large (yay, modern JS) I’ve seen the workspace persistence go slower than caching on that number of files, but not in a scientifically proven way. I’ll check it out more closely.

File: .circleci/config.yml (hunk 2 above)
Comment (Westbrook) [reply .circleci/config.yml]: I could go both ways. If we don’t block on the bs tests then it should be fine. I _do_ like that running them in parallel like this gets the whole build to less than 5 minutes pretty regularly. Confirm you want to run synchronized and I’ll update.

Comment (daKmoR) [reply .circleci/config.yml]: true let's try parallel and if at any point we want to reduce the load on browser stack we can easily do so 🤗

**Comment by daKmoR** [comment]
I think that is good to go 👍

can do one final rebase and then we can merge 👍
I will adjust the merge requirements right after 🤗

**Comment by Westbrook** [comment]
@daKmoR back up-to-date with `master`, even with Lars' approval, I don't think the repo will give me permission to merge with the branch rules not being met...

**Comment by Westbrook** [comment]
@daKmoR with so much great work happening against the repo, it's tough to keep up-to-date with `master`. PTAL.

**Comment by daKmoR** [comment]
it's merged now 👍 thxxxx 🤗

**Pull request merged**